### PR TITLE
Updated keybindings: they are now changeable.

### PR DESCRIPTION
**Pull request by lilSander**
I tried to imitate this: 
https://dota2.fandom.com/wiki/Hotkeys?file=Settings_hotkeys_basic_blank.png
The user can now change the keybindings at runtime.
When the key is already used, the user will receive an exeption text.
 I don't know for sure, but to my knowledge, we don't have a preferences file, to safe the changed keys, so the changes only last for a session.

### CODE REVIEW

File: Assets/SEE/Controls/SEEInput.cs
Comment (koschke) [inline Assets/SEE/Controls/SEEInput.cs:26]: What does this updating mean precisely?

File: Assets/SEE/Controls/SEEInput.cs
Comment (koschke) [inline Assets/SEE/Controls/SEEInput.cs:40]: The `return` is not documented.

Comment (koschke) [inline Assets/SEE/Controls/SEEInput.cs:40]: Why not `IDictionary`?

File: Assets/SEE/Controls/SEEInput.cs
Comment (koschke) [inline Assets/SEE/Controls/SEEInput.cs:60]: Do not use a string literal if it occurs more than once. Use a named string constant instead.

This should be done for all all other string literals here and in `KeyBindings.cs`. Define those string contants in `KeyBindings.cs`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:55]: Use `see cref` for all the names referenced in this comment.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:74]: Be consistent: change either to `displays` or to `instantiate`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:75]: Isn't the comment obvious from the code? Is this comment actually needed?

File: Assets/Resources/Prefabs/UI/SettingsMenu.prefab
Comment (koschke) [inline Assets/Resources/Prefabs/UI/SettingsMenu.prefab:1]: I get the following error message in the Unity Editor console:

```
Problem detected while importing the Prefab file: 'Assets/Resources/Prefabs/UI/SettingsMenu.prefab'.
The file might be corrupt or have a missing Variant parent or nested Prefabs. See details below.
Errors:
	Prefab instance problem: 'Buttons (1) (Missing Prefab with guid: 24d0ad34a12c5a44f975ec643ae33558)'. 2 instances are missing the same Prefab Asset with guid 24d0ad34a12c5a44f975ec643ae33558.
```

Do you have the same problem?

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:11]: The menu entry `Exit` must be renamed to `Exit Game` and a confirmation dialog must be added if this menu entry is selected: exiting the game is a non-reversible action and the user must be clear about that.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:105]: ESC button is no longer true.

Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:105]: `handels` => `handles`

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:111]: There must always be exactly one blank between a control keyword and the corresponding opening bracket `(`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:118]: There must always be exactly one blank between a control keyword and the corresponding opening bracket `(`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:124]: There must always be exactly one blank between a control keyword and the corresponding opening bracket `(`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:120]: There must be exactly one blank between `//` and the first word of the comment.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:160]: Add a period at the end of the comment.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:162]: Add `private`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:167]: Add `private`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:166]: Document the parameter `buttonName`. Is `buttonName` the right term to be used here?

File: Assets/SEE/UI/SettingsMenu.cs
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:170]: Remove this debugging output.

**Review by koschke — CHANGES_REQUESTED** [review]
The cheat sheet looks much better now. However, I can hardly read the descriptive text of the key bindings. The font is difficult to read. Moreover, currently the descriptive texts are internal names. A user will benefit from more understandable texts, e.g., `BoostCameraSpeed` could be `Increase Speed`. Likewise, `OpenTextChat` also closes the text chat. We can work on that later.

I found several places which need minor improvements before we can merge this PR. Please fix these. Thanks.

File: Assets/Resources/Prefabs/UI/SettingsMenu.prefab
Comment (lilSander) [reply Assets/Resources/Prefabs/UI/SettingsMenu.prefab:1]: I hope the error is gone now. Please ckeck again.

File: Assets/SEE/Controls/SEEInput.cs
Comment (lilSander) [reply Assets/SEE/Controls/SEEInput.cs:26]: There are two synchronised IDictionarys, the one in KeyBindings and another one in SEEInput, which only has the Keys and binding names. I did this, because I thought it would be a bad idea to make the bindings IDictionary in KeyBindings.cs public. But since we can change the bindings at runtime now, the IDictionary in SEEInput needs to get updated, when the user changes a key.

**Comment by koschke** [comment]
@lilSander Please review my changes.

**Comment by koschke** [comment]
Because I made so many changes as a reviewer, it would be good if someone else (maybe @falko17) could take a fresh look, to make sure my changes are reviewed by an indendent reviewer, too.

File: Assets/SEE/Controls/KeyActionDescriptor.cs
Comment (falko17) [inline Assets/SEE/Controls/KeyActionDescriptor.cs:55]: Something that I feel would be very useful here to reduce the amount of boilerplate code is a recent C# feature: [**records**](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record). This whole class can be reduced to the following, with equal functionality:

```suggestion
        /// <param name="Name">The descriptive name of the action. It will be shown to the user. It should be
        /// short and descriptive.</param>
        /// <param name="Description">A longer description of the action shown to the user. It provides more detail
        /// than <see cref="Name"/>.</param>
        /// <param name="Category">The scope of the action.</param>
        /// <param name="KeyCode">The key on the keyboard this action is bound to. If that key is pressed,
        /// the user wants to trigger this action.</param>
        internal record KeyActionDescriptor(string Name, string Description, KeyActionCategory category, KeyCode keyCode) 
        {
            /// <summary>
            /// The key on the keyboard this action is bound to. If that key is pressed,
            /// the user wants to trigger this action.
            /// </summary>
            public KeyCode KeyCode { get; set; }
        }
```

With the following semantics:
* One readonly property per record parameter, with the documentation for each field given by the parameter's documentation. We declare `KeyCode` again to make its setter accessible.
* A constructor matching the parameters.
* Human-readable `ToString()` method, similar to the current one.
* Value equality, if two instances are ever compared.

File: Assets/SEE/Controls/KeyBindings.cs
Comment (falko17) [inline Assets/SEE/Controls/KeyBindings.cs:119]: Are these intended to be displayed to the user (e.g., for reassigment)? If so, we should probably use `ShowNotification` here instead.

Comment (falko17) [inline Assets/SEE/Controls/KeyBindings.cs:119]:
```suggestion
                Debug.LogError($"Key action {keyAction} is already bound to {descriptor.Name}.\n");
```

File: Assets/SEE/Controls/KeyBindings.cs
Comment (falko17) [inline Assets/SEE/Controls/KeyBindings.cs:189]:
```suggestion
        /// <summary>
```

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:40]:
```suggestion
        /// A mapping of the short name of the key binding onto the label of the button that allows to
```

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:58]: It seems like the dictionary can be made readonly and initialized at its declaration.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:61]: Maybe this is one of those places where `var` would actually be a good idea, since the type takes up the whole line here and can be inferred pretty easily by `AllBindings().GroupBy(binding => binding.Value.Category)`.

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:86]: I think the brackets are unnecessary:
```suggestion
                    keyBindingContent.transform.Find("Key").gameObject.MustGetComponent<Button>().onClick.AddListener(() => StartRebindFor(binding.Value));
```

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:93]:
```suggestion
        /// the case that the user wants to change the key of a keyBinding.
```

File: Assets/SEE/UI/SettingsMenu.cs
Comment (falko17) [inline Assets/SEE/UI/SettingsMenu.cs:101]:
```suggestion
                // the next button that gets pressed will be the new keyBind.
```

**Review by falko17 — CHANGES_REQUESTED** [review]
The PR in general looks good, I just have a few minor suggestions.

File: Assets/SEE/Controls/KeyBindings.cs
Comment (lilSander) [inline Assets/SEE/Controls/KeyBindings.cs:186]:
```suggestion
// KeyActions from the category: General
Register(KeyAction.Help, KeyCode.H, "Help", KeyActionCategory.General, "Provides help");
Register(KeyAction.ToggleVoiceInput, KeyCode.Period, "Toggle voice chat", KeyActionCategory.General, "Toggles voice input on/off.");
Register(KeyAction.ToggleMenu, KeyCode.Space, "Toggle menu", KeyActionCategory.General, "Toggles the user-action menu.");
Register(KeyAction.ToggleSettings, KeyCode.Pause, "Toggle settings", KeyActionCategory.General, "Turns on/off the settings menu.");
Register(KeyAction.ToggleBrowser, KeyCode.F4, "Toggle browser", KeyActionCategory.General, "Turns on/off the browser.");
Register(KeyAction.SearchMenu, KeyCode.F, "Search", KeyActionCategory.General, "Opens the search menu.");
Register(KeyAction.Undo, KeyCode.Z, "Undo", KeyActionCategory.General, "Undoes the last action.");
Register(KeyAction.Redo, KeyCode.Y, "Redo", KeyActionCategory.General, "Re-does the last action.");
Register(KeyAction.ConfigMenu, KeyCode.K, "Config", KeyActionCategory.General, "Opens/closes the configuration menu.");
Register(KeyAction.TreeView, KeyCode.Tab, "Tree view", KeyActionCategory.General, "Opens/closes the tree view window.");
Register(KeyAction.CloseNotifications, KeyCode.X, "Close notifications", KeyActionCategory.General, "Clears all notifications.");
Register(KeyAction.ToggleFaceCam, KeyCode.I, "Toggle face cam", KeyActionCategory.General, "Toggles the face camera on or off.");
Register(KeyAction.ToggleFaceCamPosition, KeyCode.F3, "Toggle face-cam position", KeyActionCategory.General, "Toggles the position of the FaceCam on the player's face.");
Register(KeyAction.ToggleHolisticMetricsMenu, KeyCode.C, "Toggle metrics menu", KeyActionCategory.General, "Toggles the menu for holistic code metrics");

// KeyActions from the category: CameraPaths
Register(KeyAction.SavePathPosition, KeyCode.F11, "Save position", KeyActionCategory.CameraPaths, "Saves the current position when recording paths.");
Register(KeyAction.TogglePathPlaying, KeyCode.F12, "Toggle path playing", KeyActionCategory.CameraPaths, "Starts/stops the automated camera movement along a path.");

// KeyActions from the category: MetricCharts
Register(KeyAction.ToggleCharts, KeyCode.M, "Toggle charts", KeyActionCategory.MetricCharts, "Turns the metric charts on/off.");
Register(KeyAction.ToggleMetricHoveringSelection, KeyCode.N, "Toggle metric hovering/selection", KeyActionCategory.MetricCharts, "Toggles hovering/selection for markers in metric charts.");

// KeyActions from the category: Browsing
Register(KeyAction.ToggleEdges, KeyCode.V, "Toggle edges", KeyActionCategory.Browsing, "Toggles the visibility of all edges of a hovered code city.");
Register(KeyAction.Unselect, KeyCode.U, "Unselect", KeyActionCategory.Browsing, "Forgets all currently selected objects.");
Register(KeyAction.Cancel, KeyCode.Escape, "Cancel", KeyActionCategory.Browsing, "Cancels an action.");
Register(KeyAction.Reset, KeyCode.R, "Reset", KeyActionCategory.Browsing, "Resets a code city to its original position and scale.");
Register(KeyAction.ZoomInto, KeyCode.G, "Zoom in", KeyActionCategory.Browsing, "To zoom into a city.");
Register(KeyAction.Snap, KeyCode.LeftAlt, "Snap", KeyActionCategory.Browsing, "Snap move/rotate city.");
Register(KeyAction.DragHovered, KeyCode.LeftControl, "Drag hovered", KeyActionCategory.Browsing, "Drag code city.");
Register(KeyAction.ToggleCameraLock, KeyCode.L, "Toggle camera lock", KeyActionCategory.Browsing, "Toggles between the locked and free camera mode.");
Register(KeyAction.Pointing, KeyCode.P, "Point", KeyActionCategory.Browsing, "Toggles between Pointing.");

// KeyActions from the category: Movement
Register(KeyAction.BoostCameraSpeed, KeyCode.LeftShift, "Boost speed", KeyActionCategory.Movement, "Boosts the speed of the player movement. While pressed, movement is faster.");
Register(KeyAction.MoveForward, KeyCode.W, "Move forward", KeyActionCategory.Movement, "Move forward.");
Register(KeyAction.MoveBackward, KeyCode.S, "Move backward", KeyActionCategory.Movement, "Move backward.");
Register(KeyAction.MoveRight, KeyCode.D, "Move right", KeyActionCategory.Movement, "Move to the right.");
Register(KeyAction.MoveLeft, KeyCode.A, "Move left", KeyActionCategory.Movement, "Move to the left.");
Register(KeyAction.MoveUp, KeyCode.Q, "Move up", KeyActionCategory.Movement, "Move up.");
Register(KeyAction.MoveDown, KeyCode.E, "Move down", KeyActionCategory.Movement, "Move down.");

// KeyActions from the category: Evolution
Register(KeyAction.SetMarker, KeyCode.Insert, "Set marker", KeyActionCategory.Evolution, "Sets a new marker.");
Register(KeyAction.DeleteMarker, KeyCode.Delete, "Delete marker", KeyActionCategory.Evolution, "Deletes a marker.");
Register(KeyAction.ToggleEvolutionCanvases, KeyCode.T, "Toggle evolution canvases", KeyActionCategory.Evolution, "Toggles between between the two canvases for the animation and selection of a revision.");

// KeyActions from the category: Animation
Register(KeyAction.Previous, KeyCode.LeftArrow, "Previous", KeyActionCategory.Animation, "Go to previous element in the animation.");
Register(KeyAction.Next, KeyCode.RightArrow, "Next", KeyActionCategory.Animation, "Go to next element in the animation.");
Register(KeyAction.ToggleAutoPlay, KeyCode.F9, "Toggle auto play", KeyActionCategory.Animation, "Toggles auto play of the animation.");
Register(KeyAction.IncreaseAnimationSpeed, KeyCode.UpArrow, "Increase animation speed", KeyActionCategory.Animation, "Doubles animation speed.");
Register(KeyAction.DecreaseAnimationSpeed, KeyCode.DownArrow, "Decrease animation speed", KeyActionCategory.Animation, "Halves animation speed.");

// KeyActions from the category: Debugging
Register(KeyAction.ToggleExecutionOrder, KeyCode.O, "Toggle execution order", KeyActionCategory.Debugging, "Toggles execution order (foward/backward).");
Register(KeyAction.ExecuteToBreakpoint, KeyCode.B, "Execute to breakpoint", KeyActionCategory.Debugging, "Continues execution until next breakpoint is reached.");
Register(KeyAction.FirstStatement, KeyCode.Home, "First statement", KeyActionCategory.Debugging, "Execution is back to very first statement.");

// KeyActions from the category: CodeViewer
Register(KeyAction.ShowWindowMenu, KeyCode.F1, "Show window menu", KeyActionCategory.CodeViewer, "Toggles the menu of the open windows.");
Register(KeyAction.CodeWindowUndo, KeyCode.F5, "Code-window undo", KeyActionCategory.CodeViewer, "Undoes an edit in the source-code viewer.");
Register(KeyAction.CodeWindowRedo, KeyCode.F6, "Code-window redo", KeyActionCategory.CodeViewer, "Redoes an undone edit in the source-code viewer.");
Register(KeyAction.CodeWindowSave, KeyCode.F7, "Code-window save", KeyActionCategory.CodeViewer, "Saves the content of the source-code viewer.");
Register(KeyAction.RefreshSyntaxHighlighting, KeyCode.F8, "Refresh", KeyActionCategory.CodeViewer, "Refreshes syntax highlighting in the source-code viewer.");

// KeyActions from the category: Chat
Register(KeyAction.ToggleTextChat, KeyCode.F2, "Toggle text chat", KeyActionCategory.Chat, "Toggles the text chat.");
```

File: Assets/SEE/Controls/KeyBindings.cs
Comment (lilSander) [reply Assets/SEE/Controls/KeyBindings.cs:134]: And not all KeyActions were grouped together.

**Review by lilSander — COMMENTED** [review]
Overall, your changes look good to me. I only found one thing which could be changed.
To make it easy to add/delete/change keyBindings in the future, we should separate them in there categories to keep a better  overview.

File: Assets/SEE/Controls/KeyActionDescriptor.cs
Comment (koschke) [reply Assets/SEE/Controls/KeyActionDescriptor.cs:55]: We need to be able to modify the `keyCode` field. In case of a `struct`, all its fields are read-only.

File: Assets/SEE/Controls/KeyBindings.cs
Comment (koschke) [reply Assets/SEE/Controls/KeyBindings.cs]: I fixed it.

**Review by koschke — APPROVED** [review]
All looks good to me.

File: Assets/SEE/Controls/KeyActionDescriptor.cs
Comment (falko17) [reply Assets/SEE/Controls/KeyActionDescriptor.cs:55]: An error I made in my original suggestion lead to a misunderstanding – the following should work as intended, including reference semantics (`record` as shorthand for `record class`) and a modifiable `KeyCode` field.
```suggestion
        /// <param name="Name">The descriptive name of the action. It will be shown to the user. It should be
        /// short and descriptive.</param>
        /// <param name="Description">A longer description of the action shown to the user. It provides more detail
        /// than <see cref="Name"/>.</param>
        /// <param name="Category">The scope of the action.</param>
        /// <param name="KeyCode">The key on the keyboard this action is bound to. If that key is pressed,
        /// the user wants to trigger this action.</param>
        internal record KeyActionDescriptor(string Name, string Description, KeyActionCategory Category, KeyCode KeyCode) 
        {
            /// <summary>
            /// The key on the keyboard this action is bound to. If that key is pressed,
            /// the user wants to trigger this action.
            /// </summary>
            public KeyCode KeyCode { get; set; }
        }
```

**Review by falko17 — APPROVED** [review]

**Review by koschke — CHANGES_REQUESTED** [review]
The replacement as a `record` for `KeyActionDescriptor` does not work. Keyboard input is ignored. Visual Studio complains that parameter/record field `KeyCode` is unused.

**Review by koschke — APPROVED** [review]

**Comment by falko17** [comment]
> The replacement as a `record` for `KeyActionDescriptor` does not work. Keyboard input is ignored. Visual Studio complains that parameter/record field `KeyCode` is unused.

Alright, I'll try out later what's going wrong here locally. To not hold up the merge any longer, we should just merge with the original `class` version, I think.

**Pull request merged**
